### PR TITLE
Feature better linting

### PR DESCRIPTION
**Pull request by irm-codebase**
#607 / #611 mostly happened because inputs and docstrings were not being evaluated by the linter.

I propose to activate additional linting. Docstring linting tends to be annoying, but in our case I think it'll help.
It's very likely that this will trigger some hassle when modifying code at the start, but in the long run it will make our lives easier.

Additional linting is not always well received, though. Please let me know what you think!

## Summary of changes in this pull request

*  I am activating two additional tests: "D" ([docstyle](https://docs.astral.sh/ruff/rules/#pydocstyle-d)) and "PT" ([pytest-style](https://docs.astral.sh/ruff/rules/#flake8-pytest-style-pt)).
* Added editor config for additional consistency in .YAML files

## Reviewer checklist

- [ ] Test(s) added to cover contribution
- [ ] Documentation updated
- [x] Changelog updated
- [ ] Coverage maintained or improved

### CODE REVIEW

**Comment by irm-codebase** [comment]
`pre-commit` breaking is expected. I will fix these issues before merging this PR, but only after confirming these new lints are desired.

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+end_of_line = lf
+insert_final_newline = true
+charset = utf-8
+
+[*.{yaml,yml}]
+indent_size = 2
+
+[*.md]
+trim_trailing_whitespace = true
+indent_size = 4
```
Comment (brynpickering) [inline .editorconfig:19]: how does this work when doing two-line indenting inside yaml code blocks in the docs?

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+[LICENSE]
+insert_final_newline = false
+
+[*.{diff,patch}]
```
Comment (brynpickering) [inline .editorconfig:24]: are these files even tracked? I assume they'd all be in the `.gitignore`

**Review by brynpickering — COMMENTED** [review]
If you're happy to fix the issues raised by `precommit.ci` then I'm strongly in favour of this!

File: .editorconfig (hunk 1 above)
Comment (irm-codebase) [reply .editorconfig:19]: It will still apply 4. I'll update `.md` files to use 2 by default. Honestly not that big of a problem since most markdown previews do not need the 4 spaces.

File: .editorconfig (hunk 2 above)
Comment (irm-codebase) [reply .editorconfig:24]: They are not in the `gitingore`... but we do not even use them. I'll just remove this.

**Comment by irm-codebase** [comment]
> If you're happy to fix the issues raised by `precommit.ci` then I'm strongly in favour of this!
Happy is a _strong_ word for linting fixes... but I will do it hahaha.

File: .editorconfig (hunk 1 above)
Comment (brynpickering) [reply .editorconfig:19]: no, that won't work either. 4 lines is required for some mkdocs features to work

Comment (irm-codebase) [reply .editorconfig:19]: .editorconfig is not automatic in that sense. It only alters the default behavior of pressing "tab" in the editor. It is not going to modify files.

There are 2 options:
- Setting 4 then modifying to 2.
- Setting 2 and being aware of the 4 cases.

Let's go with whichever is more generic.

Comment (brynpickering) [reply .editorconfig:19]: Ok, go with 4

**Comment by brynpickering** [comment]
@irm-codebase don't forget to add yourself to the list in `AUTHORS` during this PR

**Comment by irm-codebase** [comment]
@brynpickering good chuck of files done. Most of these are minor changes.
I am not completely re-working the documentation (I think it's better/safer to do that as we go on).

Question though: is it possible to configure pre-commit to only run "D" (pydocstrings) for merges?
I think it makes sense to allow people to develop their code with docstring flexiblity, and only enforce it before bringing it into main.

I've read a bit of pre-commit's configuration, and it seems possible, but I am not experienced enough with setting hooks to know how to achieve it.

**Comment by irm-codebase** [comment]
This additional setting in `pr-ci.yml` could do the trick, but it depends on whether or not we want people to be pestered by docstrings during regular commits, or have them be surprised by extra changes in the PR.

```yaml
  ruff-docstring:
    runs-on: ubuntu-latest
    if: startsWith(github.event.pull_request.title, 'Release v')
    steps:
      - uses: chartboost/ruff-action@v1
        with:
          args: check --select D
```

**Comment by brynpickering** [comment]
I think I prefer it being there by default. Hopefully shouldn't be too much effort to keep pre-commit / ruff happy with additions once you have the fixes done in this PR. It instills good documenting practice from our development team, which is no bad thing ;)

File: .pre-commit-config.yaml
Comment (irm-codebase) [inline .pre-commit-config.yaml:1]: As requested, no more black. I also updated the version of Ruff in pre-commit.

File: src/calliope/backend/helper_functions.py
Comment (irm-codebase) [inline src/calliope/backend/helper_functions.py:1]: Please go over this file. I am a tad worried about my lack of familiarity with the parser, and the naming was a bit hard to follow.

File: pyproject.toml
Comment (irm-codebase) [inline pyproject.toml:1]: Removed black and replaced with ruff format.
I only included checks for files not in the Ruff defaults, to keep the file lean.

I also _removed_ the magic comma skipping because it conflicted with our current isort (I) configuration!

File: pyproject.toml [hunk 3]
```diff
@@ -30,56 +30,40 @@ directory = "reports/coverage"
...
 # Unlike Flake8, default to a complexity level of 10.
 max-complexity = 10
 
 # Ignore `E402` (import violations) and `F401` (unused imports) in all `__init__.py` files
-[tool.ruff.per-file-ignores]
+[tool.ruff.lint.per-file-ignores]
 "__init__.py" = ["E402", "F401"]
 "*.ipynb" = ["E402"]
+"tests/*" = ["D"]
+"docs/examples/*" = ["D"]
```
Comment (irm-codebase) [inline pyproject.toml:53]: I am skipping the docs because the added markdown makes docstrings redundant.

In the case of tests, because he have more than a thousand and the tester should be the one adding it!
I suggest that, from now on, every new/modified test should have a short one liner describing its goal with an imperative.

e.g.`"""Sorting must be lower to upper by default"""`

**Review by irm-codebase — COMMENTED** [review]
Ready for review @brynpickering !

Again, this is just a start to good documentation. I tried to be light on what I added because I was not the main developer. If you have the time, put special attention to my docstrings in the parser functionality. That's the only one I am worried about.

File: src/calliope/backend/helper_functions.py [hunk 4]
```diff
@@ -24,6 +23,8 @@
 
 
 class ParsingHelperFunction(ABC):
+    """Helper class for parsing."""
```
Comment (brynpickering) [inline src/calliope/backend/helper_functions.py:26]:
```suggestion
    """Abstract base class for helper function parsing."""
```

File: src/calliope/backend/latex_backend_model.py [hunk 5]
```diff
@@ -362,7 +374,7 @@ def add_parameter(
         default: Any = np.nan,
         use_inf_as_na: bool = False,
     ) -> None:
-
+        """Add parameter to docs."""
```
Comment (brynpickering) [inline src/calliope/backend/latex_backend_model.py:377]: These (and below) shouldn't be needed. They inherit their docs from the abstract base class. By adding this line in, I expect the inherited docsstrings to be overwritten, which is not ideal as those are quite verbose. If ruff can't handle this then we need some way to deal with it that doesn't involve overwriting what's in the ABC

File: src/calliope/backend/parsing.py [hunk 6]
```diff
@@ -762,21 +789,26 @@ def generate_top_level_where_array(
         *,
         align_to_foreach_sets: bool = True,
         break_early: bool = True,
-        references: Optional[set] = None,
+        references: set | None = None,
```
Comment (brynpickering) [inline src/calliope/backend/parsing.py:792]: Is it now recommended to _not_ use `Optional`?

File: src/calliope/backend/where_parser.py [hunk 7]
```diff
@@ -282,42 +304,49 @@ def __init__(self, instring: str, loc: int, tokens: pp.ParseResults) -> None:
...
         return f"SUBSET:{self.set_name}{self.val}"
 
     def _eval(self) -> list[str | float]:
-        "Evaluate each element of the subset list"
+        """Evaluate each element of the subset list."""
         values = [val.eval("array", **self.eval_attrs) for val in self.val]
         return [val.item() if isinstance(val, xr.DataArray) else val for val in values]
 
     def as_math_string(self) -> str:
+        """Process and return as LaTeX."""
```
Comment (brynpickering) [inline src/calliope/backend/where_parser.py:316]: As with the backend models, these docstrings are inherited from the ABC so they shouldn't need to be repeated ad nauseam. We need a solution for this on the linting side

File: src/calliope/preprocess/load.py [hunk 8]
```diff
@@ -42,12 +34,13 @@ def load_model_definition(
...
             These will be applied _after_ `scenario` overrides.
             Defaults to None.
+        **kwargs: initialisation overrides.
 
     Returns:
         tuple[AttrDict, Optional[Path], str]:
-            1. Model definition with overrides applied.
-            1. Path to model definition YAML if input `model_definiton` was pathlike, otherwise None.
-            1. Expansion of scenarios (which are references to model overrides) into a list of named override(s) that have been applied.
+            - Model definition with overrides applied.
```
Comment (brynpickering) [inline src/calliope/preprocess/load.py:41]: I previously had it numbered so that it would be obvious the order of items in the tuple, is bullet points recommended instead?

**Review by brynpickering — COMMENTED** [review]
Thanks for the effort @irm-codebase !

RE helper functions, we have a bit more detail on each [in the docs](https://calliope.readthedocs.io/en/latest/user_defined_math/syntax/#helper-functions). Could you take snippets from there to use for the introduction docstring for each class?

Then the only major thing is the requirement for docstrings in overrides to abstract methods. We really don't want to have to repeat a docstring that is already given in full by the ABC.

File: src/calliope/preprocess/load.py (hunk 8 above)
Comment (irm-codebase) [reply src/calliope/preprocess/load.py:41]: Oh, I'll re introduce them. They all had 1 so I thought it was fine.

File: src/calliope/backend/where_parser.py (hunk 7 above)
Comment (irm-codebase) [reply src/calliope/backend/where_parser.py:316]: Gah, you are right. I wish we could just use `@override` (https://docs.python.org/3/library/typing.html#typing.override), then the linter would catch these cases.

Unfortunately it would break compatibility with 3.11 and lower :(

Comment (irm-codebase) [reply src/calliope/backend/where_parser.py:316]: I'll just add   # noqa: D102, override. Already did it for other functions, it'll be enough for now.

**Comment by irm-codebase** [comment]
@brynpickering corrections implemented!
Sadly the only way to avoid the `@abstractmethod` issue is to deactivate linting for those functions.
We could implement the `@override` decorator to avoid this, but it's python 3.12 exclusive, and will break our tests for 3.10 / 3.11.

**Comment by irm-codebase** [comment]
No idea why codeconv does not like this update, there is nothing added to the code :shrug:

**Comment by brynpickering** [comment]
> edit: it's because of the added ... in some @abstractmethod cases. But without it we get linter issues.

Yeah, I think that's why I had docsstrings in them to begin with, to keep codecov happy. I'm not sure there's much to be done about it (or, at least, I can't find an obvious answer online).

File: src/calliope/backend/helper_functions.py
Comment (brynpickering) [reply src/calliope/backend/helper_functions.py:1]: Updated

**Comment by irm-codebase** [comment]
Thanks for the help @brynpickering! I went through your changes and I agree to them.
Is there anything else left to do for this PR?

**Review by brynpickering — APPROVED** [review]
